docs/design-proposals/2021-11-24-graceful-reschedule-tikv-pod.md

```diff
@@ -0,0 +1,164 @@
+2. When the leader count drops down to zero, delete the pod to let it re-create and reschedule.
+3. Remove the evict-leader-scheduler.
+
+## Motivation
+
+### Goals
+
+- Provide a way for a user to gracefully reschedule a tikv pod.
+- The pod must be gracefully stopped, to be specified, evict leaders before deleting the pod, and must work normally after recreating the pod.
+
+### Non-Goals
+
+- Provider a way to make tikv graceful-stop when just receiving a `SIGTERM` signal.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+suppose `tikv-0` is running at node `node0`,  there are many free resources at other nodes, I would like to make `tikv-0` run at other node and shutdown `node0` to reduce the cost. I may apply the flowing operations:
+
+1. run `kubectl cordon node0` to mark `node0` as `unschedulable`.
+2. delete pod `tikv-0` to let it re-create and reschedule to other nodes, to mitigate the impact of unavailability of some regions, I will do the flowing operations instead of deleting the pod directly:
+   1. add evict-leader-scheduler by `pd-ctl`.
+   2. delete `tikv-0` pod when leader count drops down to 0.
+   3. remove the evict-leader-scheduler by `pd-ctl`
+3. drain `node-0` in the normal way as [safely drain a node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) and shutdown `node0`
+
+### Risks and Mitigations
+
+- PD/TiKV fails to evict all leaders and we may just keep checking it.
+
+## Design Details
+
+Support user to add an annotation to tikv pod to trigger a graceful reschedule.
+
+Annotation key: `tidb.pingcap.com/restart`
+
+The controller will do the flowing operations:
+
+1. Add evict-leader-scheduler for the tikv store.
+2. Delete to the pod to make it re-create when leader-count is 0.
+3. Remove the evict-leader-scheduler when the new pod becomes ready.
+
+
+
+For the user only want to trigger an evict-leader-scheduler by annotating a pod, we will also support another key for this.
+
+Annotation key:`tidb.pingcap.com/evict-leader`
+
+The controller will do the flowing operations:
+
+1. Add evict-leader-scheduler for the tikv store.
+
+2. Remove the evict-leader-scheduler once is annotation is deleted.
+
+
+
+An EvictLeader status will be added to the `TiKVStatus`:
+
+```go
++type EvictLeaderStatus struct {
++       PodCreateTime metav1.Time `json:"podCreateTime,omitempty"`
++       Value         string      `json:"value,omitempty"`
++}
+
+ // TiKVStatus is TiKV status
+ type TiKVStatus struct {
+        Synced          bool                            `json:"synced,omitempty"`
+@@ -1139,6 +1151,7 @@ type TiKVStatus struct {
+        TombstoneStores map[string]TiKVStore            `json:"tombstoneStores,omitempty"`
+        FailureStores   map[string]TiKVFailureStore     `json:"failureStores,omitempty"`
+        Image           string                          `json:"image,omitempty"`
++       EvictLeader     map[string]*EvictLeaderStatus   `json:"evictLeaderStatus,omitempty"`
```

Comment:
Is `EvictLeaderStatus` useful here? If not, can we just use `EvictLeaderPods map[string]EmptyStruct `json:"evictLeaderPods,omitempty"``?